meetupselector/api/schemas/users.py

```diff
@@ -0,0 +1,22 @@
+import re
+
+from pydantic import BaseModel, EmailStr, validator
+
+
+class SignInSchema(BaseModel):
+    email: EmailStr
+    password: str
+
+    @validator("password")
+    def password_validator(cls, value: str) -> str:
+        msg_error = (
+            "password requirements: min. length of 8 chars, "
+            "one uppercase char, one lowercase char, one digit, "
+            "one special char (not letter, not number)"
+        )
+        patterns = (".{8}", "[A-Z]", "[a-z]", "[0-9]", "[^a-zA-Z0-9]")
+        for pattern in patterns:
+            if re.search(pattern, value) is None:
+                raise ValueError(msg_error)
+
+        return value
```

Comment:
I would delegate password validation to Django. You can use [`validate_password`](https://docs.djangoproject.com/en/4.1/topics/auth/passwords/#django.contrib.auth.password_validation.validate_password) to do it.

Reply:
You're right and it was my first approach. [According to doc](https://docs.djangoproject.com/en/4.1/topics/auth/passwords/#module-django.contrib.auth.password_validation), the steps I did:

- Write 4 custom validators. One per each password requirements: at least one upper letter, at least one lower letter, at least one digit and at least one special char.
- Write a test suite per each of the 4 validators ([see doc here](https://docs.djangoproject.com/en/4.1/topics/auth/passwords/#writing-your-own-validator)). Note: `min_length` validator is already built-in in django.
- Modify settings in order to add those validators to the list of validators Django will use ([see doc here](https://docs.djangoproject.com/en/4.1/topics/auth/passwords/#enabling-password-validation)).
- Use `validate_password` in our application service.

In addition to the above, we also have to validate email. And Django Ninja is not validating it, even having defined Django model as EmailField. So, sending any string in e-mail field was passing validation without errors end being saved in DB :(

But then, I realized that would be much easier to use [Pydantic validators](https://pydantic-docs.helpmanual.io/usage/validators/). Email validator is already built-in in Pydantic and writing a validator for password takes only few lines of code.

Reply:
I don't know why `models.EmailField` doesn't validate the email, because in the admin panel it is validated.
![admin-backend-wrong-email](https://user-images.githubusercontent.com/2360683/199336594-eab27b61-b8a7-4052-a4dd-09c3486e080c.png)

But in User model `create_user` it isn't:
```python
>>> from meetupselector.user.models import User
>>> User.objects.create_user(email="wrong", password="V4l!dP4sSw0rD!")
<User: wrong>
```

Reply:
Regarding passwords, I only have one comment:
![image](https://user-images.githubusercontent.com/2360683/199337995-4a88ac04-b029-438a-bc8e-23880e3e385b.png)
[source](https://xkcd.com/936/)

Jokes aside, I propose to implement two validators, one as you've already done, and if it doesn't pass, to check also if the password is a passphrase.

Reply:
> But then, I realized that would be much easier to use [Pydantic validators](https://pydantic-docs.helpmanual.io/usage/validators/). Email validator is already built-in in Pydantic and writing a validator for password takes only few lines of code.

Do you think that password and email validations should be in api layer? Or in domain layer?

Reply:
> Write 4 custom validators. One per each password requirements: at least one upper letter, at least one lower letter, at least one digit and at least one special char. Note: min_length validator is already built-in in django, so I didn't write a validator for that. ([see doc here](https://docs.djangoproject.com/en/4.1/topics/auth/passwords/#writing-your-own-validator)).
    Write a test suite per each of the 4 validators.
    Modify settings in order to add those validators to the list of validators Django will use ([see doc here](https://docs.djangoproject.com/en/4.1/topics/auth/passwords/#enabling-password-validation)).
    Use validate_password in our application service.
    
In my opinion this was the right approach because it allows us to have validators in the model. I've searched and for some reason django ninja don't apply model validators in the schema... but we can do it ourselves.

Something like this:
```python
from django.contrib.auth.password_validation import validate_password

class SignInSchema(BaseModel):
    email: EmailStr
    password: str

    @validator("password")
    def password_validator(cls, value: str) -> str:
        try:
            validate_password(value)
        except ValidationError as error:
            message = ...  # TODO: grab message from error
            raise ValueError(message)
        
```

As for the email I see it less problematic to use pydantic validator, but we could also do a validator with `validate_email`

Reply:
> In my opinion this was the right approach because it allows us to have validators in the model. 

@serfer2 @EnriqueSoria 

Django doesn't validate the email or password unless you call it specifically. See my previous comment about this topic. It only performs validation on Django Forms.

By the way, we can call the validators in `User.objects.create_user` method.